--- a/apps/CA.pl.in
+++ b/apps/CA.pl.in
@@ -29,7 +29,7 @@
 my $PKCS12 = "$openssl pkcs12";

 # Default values for various configuration settings.
-my $CATOP = "./demoCA";
+my $CATOP = "/etc/ssl";
 my $CAKEY = "cakey.pem";
 my $CAREQ = "careq.pem";
 my $CACERT = "cacert.pem";
--- a/apps/openssl.cnf
+++ b/apps/openssl.cnf
@@ -79,7 +79,7 @@
 ####################################################################
 [ CA_default ]

-dir		= ./demoCA		# Where everything is kept
+dir		= /etc/ssl		# Where everything is kept
 certs		= $dir/certs		# Where the issued certs are kept
 crl_dir		= $dir/crl		# Where the issued crl are kept
 database	= $dir/index.txt	# database index file.
@@ -309,7 +309,7 @@
 [ tsa_config1 ]

 # These are used by the TSA reply generation only.
-dir		= ./demoCA		# TSA root directory
+dir		= /etc/ssl		# TSA root directory
 serial		= $dir/tsaserial	# The current serial number (mandatory)
 crypto_device	= builtin		# OpenSSL engine to use for signing
 signer_cert	= $dir/tsacert.pem 	# The TSA signing certificate

From: Auke Kok <auke-jan.h.kok@intel.com>
Date: Mon, 22 Oct 2018 13:27:06 -0700
Subject: [PATCH] Use clearlinux CFLAGS during build.

---
 Configurations/10-main.conf | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/Configurations/10-main.conf b/Configurations/10-main.conf
index 8dc3e85..67470e9 100644
--- a/Configurations/10-main.conf
+++ b/Configurations/10-main.conf
@@ -657,6 +657,7 @@ my %targets = (
     },
     "linux-generic64" => {
         inherit_from     => [ "linux-generic32" ],
+        CFLAGS           => add($ENV{'CFLAGS'}),
         bn_ops           => "SIXTY_FOUR_BIT_LONG RC4_CHAR",
     },
