diff --git a/test/meson.build b/test/meson.build
index 79fce47..c5ec184 100644
--- a/test/meson.build
+++ b/test/meson.build
@@ -198,7 +198,7 @@ if build_xorg
 
     if meson.get_compiler('c').has_link_argument('-Wl,-wrap')
        # LTO breaks with -Wl,-wrap on certain configurations
-       unit_c_args += ['-fno-lto']
+       unit_c_args += ['-flto']
        unit_sources += [
         'xi1/protocol-xchangedevicecontrol.c',
         'xi2/protocol-common.c',
