--- renderloop.h.orig	2025-10-06 01:22:45.992267086 +0200
+++ renderloop.h	2025-10-06 12:13:11.625191349 +0200
@@ -9,6 +9,9 @@
 #include "effect/globals.h"
 
 #include <QObject>
+#include <chrono>
+#include <memory>
+#include <optional>
 
 namespace KWin
 {
@@ -51,8 +54,7 @@ public:
     void uninhibit();
 
     /**
-     * This function must be called before the Compositor sumbits the next
-     * frame.
+     * This function must be called before the Compositor submits the next frame.
      */
     void prepareNewFrame();
 
@@ -95,26 +97,11 @@ public:
      */
     std::chrono::nanoseconds predictedRenderTime() const;
 
-    // TODO integrate cursor updates into the render loop / frame scheduling somehow?
-    // and then remove this again
     bool activeWindowControlsVrrRefreshRate() const;
 
 Q_SIGNALS:
-    /**
-     * This signal is emitted when the refresh rate of this RenderLoop has changed.
-     */
     void refreshRateChanged();
-    /**
-     * This signal is emitted when a frame has been actually presented on the screen.
-     * @a timestamp indicates the time when it took place.
-     */
     void framePresented(RenderLoop *loop, std::chrono::nanoseconds timestamp, PresentationMode mode);
-
-    /**
-     * This signal is emitted when the render loop wants a new frame to be composited.
-     *
-     * The Compositor should make a connection to this signal using Qt::DirectConnection.
-     */
     void frameRequested(RenderLoop *loop);
 
 private:
