--- surfaceitem.cpp.orig	2025-10-06 00:05:53.790543653 +0200
+++ surfaceitem.cpp	2025-10-06 00:29:54.039175947 +0200
@@ -105,10 +105,11 @@ QRegion SurfaceItem::mapFromBuffer(const
     const QRectF sourceBox = m_bufferToSurfaceTransform.map(m_bufferSourceBox, m_bufferSize);
     const qreal xScale = m_destinationSize.width() / sourceBox.width();
     const qreal yScale = m_destinationSize.height() / sourceBox.height();
+    const QPointF sourceBoxTopLeft = sourceBox.topLeft();
 
     QRegion result;
-    for (QRectF rect : region) {
-        const QRectF r = m_bufferToSurfaceTransform.map(rect, m_bufferSize).translated(-sourceBox.topLeft());
+    for (const QRect &rect : region) {
+        const QRectF r = m_bufferToSurfaceTransform.map(QRectF(rect), m_bufferSize).translated(-sourceBoxTopLeft);
         result += QRectF(r.x() * xScale, r.y() * yScale, r.width() * xScale, r.height() * yScale).toAlignedRect();
     }
     return result;
@@ -120,11 +121,15 @@ static QRegion expandRegion(const QRegio
         return QRegion();
     }
 
-    QRegion ret;
+    QVector<QRect> expandedRects;
+    expandedRects.reserve(region.rectCount());
+
     for (const QRect &rect : region) {
-        ret += rect.marginsAdded(padding);
+        expandedRects.append(rect.marginsAdded(padding));
     }
 
+    QRegion ret;
+    ret.setRects(expandedRects.constData(), expandedRects.size());
     return ret;
 }
 
@@ -136,7 +141,7 @@ void SurfaceItem::addDamage(const QRegio
         if (m_lastDamageTimeDiffs.size() > 100) {
             m_lastDamageTimeDiffs.pop_front();
         }
-        m_frameTimeEstimation = std::accumulate(m_lastDamageTimeDiffs.begin(), m_lastDamageTimeDiffs.end(), 0ns) / m_lastDamageTimeDiffs.size();
+        m_frameTimeEstimation = std::accumulate(m_lastDamageTimeDiffs.begin(), m_lastDamageTimeDiffs.end(), 0ns) / static_cast<std::chrono::nanoseconds::rep>(m_lastDamageTimeDiffs.size());
     }
     m_lastDamage = std::chrono::steady_clock::now();
     m_damage += region;
@@ -222,14 +227,16 @@ WindowQuadList SurfaceItem::buildQuads()
     const QRectF sourceBox = m_bufferToSurfaceTransform.map(m_bufferSourceBox, m_bufferSize);
     const qreal xScale = sourceBox.width() / m_destinationSize.width();
     const qreal yScale = sourceBox.height() / m_destinationSize.height();
+    const QPointF bufferSourceTopLeft = m_bufferSourceBox.topLeft();
+    const QSizeF sourceBoxSize = sourceBox.size();
 
-    for (const QRectF rect : region) {
+    for (const QRectF &rect : region) {
         WindowQuad quad;
 
-        const QPointF bufferTopLeft = snapToPixelGridF(m_bufferSourceBox.topLeft() + m_surfaceToBufferTransform.map(QPointF(rect.left() * xScale, rect.top() * yScale), sourceBox.size()));
-        const QPointF bufferTopRight = snapToPixelGridF(m_bufferSourceBox.topLeft() + m_surfaceToBufferTransform.map(QPointF(rect.right() * xScale, rect.top() * yScale), sourceBox.size()));
-        const QPointF bufferBottomRight = snapToPixelGridF(m_bufferSourceBox.topLeft() + m_surfaceToBufferTransform.map(QPointF(rect.right() * xScale, rect.bottom() * yScale), sourceBox.size()));
-        const QPointF bufferBottomLeft = snapToPixelGridF(m_bufferSourceBox.topLeft() + m_surfaceToBufferTransform.map(QPointF(rect.left() * xScale, rect.bottom() * yScale), sourceBox.size()));
+        const QPointF bufferTopLeft = snapToPixelGridF(bufferSourceTopLeft + m_surfaceToBufferTransform.map(QPointF(rect.left() * xScale, rect.top() * yScale), sourceBoxSize));
+        const QPointF bufferTopRight = snapToPixelGridF(bufferSourceTopLeft + m_surfaceToBufferTransform.map(QPointF(rect.right() * xScale, rect.top() * yScale), sourceBoxSize));
+        const QPointF bufferBottomRight = snapToPixelGridF(bufferSourceTopLeft + m_surfaceToBufferTransform.map(QPointF(rect.right() * xScale, rect.bottom() * yScale), sourceBoxSize));
+        const QPointF bufferBottomLeft = snapToPixelGridF(bufferSourceTopLeft + m_surfaceToBufferTransform.map(QPointF(rect.left() * xScale, rect.bottom() * yScale), sourceBoxSize));
 
         quad[0] = WindowVertex(rect.topLeft(), bufferTopLeft);
         quad[1] = WindowVertex(rect.topRight(), bufferTopRight);
@@ -357,11 +364,11 @@ OpenGLSurfaceContents OpenGLSurfaceTextu
 bool OpenGLSurfaceTexture::create()
 {
     GraphicsBuffer *buffer = m_pixmap->item()->buffer();
-    if (buffer->dmabufAttributes()) {
+    if (Q_LIKELY(buffer->dmabufAttributes())) {
         return loadDmabufTexture(buffer);
     } else if (buffer->shmAttributes()) {
         return loadShmTexture(buffer);
-    } else if (buffer->singlePixelAttributes()) {
+    } else if (Q_UNLIKELY(buffer->singlePixelAttributes())) {
         return loadSinglePixelTexture(buffer);
     } else {
         qCDebug(KWIN_OPENGL) << "Failed to create OpenGLSurfaceTexture for a buffer of unknown type" << buffer;
@@ -378,11 +385,11 @@ void OpenGLSurfaceTexture::destroy()
 void OpenGLSurfaceTexture::update(const QRegion &region)
 {
     GraphicsBuffer *buffer = m_pixmap->item()->buffer();
-    if (buffer->dmabufAttributes()) {
+    if (Q_LIKELY(buffer->dmabufAttributes())) {
         updateDmabufTexture(buffer);
     } else if (buffer->shmAttributes()) {
         updateShmTexture(buffer, region);
-    } else if (buffer->singlePixelAttributes()) {
+    } else if (Q_UNLIKELY(buffer->singlePixelAttributes())) {
         updateSinglePixelTexture(buffer);
     } else {
         qCDebug(KWIN_OPENGL) << "Failed to update OpenGLSurfaceTexture for a buffer of unknown type" << buffer;
@@ -414,7 +421,7 @@ bool OpenGLSurfaceTexture::loadShmTextur
 
 static QRegion simplifyDamage(const QRegion &damage)
 {
-    if (damage.rectCount() < 3) {
+    if (damage.rectCount() < 8) {
         return damage;
     } else {
         return damage.boundingRect();
