--- wayland_backend.h.orig	2025-10-06 12:34:15.532162274 +0200
+++ wayland_backend.h	2025-10-06 13:58:36.524874704 +0200
@@ -17,6 +17,7 @@
 #include "core/outputbackend.h"
 #include "effect/globals.h"
 #include "utils/filedescriptor.h"
+
 // Qt
 #include <QHash>
 #include <QImage>
@@ -24,6 +25,11 @@
 #include <QPoint>
 #include <QSize>
 
+// STL
+#include <unordered_map>
+#include <memory>
+#include <map>
+
 struct wl_buffer;
 struct wl_display;
 
@@ -263,7 +269,7 @@ private:
     bool m_pointerLockRequested = false;
     std::unique_ptr<DrmDevice> m_drmDevice;
     std::unique_ptr<EglDisplay> m_eglDisplay;
-    std::map<GraphicsBuffer *, std::unique_ptr<WaylandBuffer>> m_buffers;
+    std::unordered_map<GraphicsBuffer *, std::unique_ptr<WaylandBuffer>> m_buffers;
 };
 
 } // namespace Wayland
