From 80247e1f62f64f6f2c1b9f2f7e6321fc66945d4d Mon Sep 17 00:00:00 2001
From: Twaik Yont <twaikyont@gmail.com>
Date: Fri, 10 Mar 2023 16:10:46 +0200
Subject: [PATCH] xvfb: Use RROutputSetPhysicalSize to set physical size of
 display

Signed-off-by: Twaik Yont <twaikyont@gmail.com>
Closes #1532
---
 hw/vfb/InitOutput.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/hw/vfb/InitOutput.c b/hw/vfb/InitOutput.c
index 48efb61b2f..ae9906c588 100644
--- a/hw/vfb/InitOutput.c
+++ b/hw/vfb/InitOutput.c
@@ -753,6 +753,7 @@ vfbRRScreenSetSize(ScreenPtr  pScreen,
                    CARD32     mmWidth,
                    CARD32     mmHeight)
 {
+    rrScrPrivPtr pScrPriv = rrGetScrPriv(pScreen);
     // Prevent screen updates while we change things around
     SetRootClip(pScreen, ROOT_CLIP_NONE);
 
@@ -767,7 +768,7 @@ vfbRRScreenSetSize(ScreenPtr  pScreen,
     RRScreenSizeNotify (pScreen);
     RRTellChanged(pScreen);
 
-    return TRUE;
+    return RROutputSetPhysicalSize(pScrPriv->outputs[pScreen->myNum], mmWidth, mmHeight);
 }
 
 static Bool
@@ -803,6 +804,7 @@ vfbRandRInit(ScreenPtr pScreen)
     xRRModeInfo modeInfo;
     char       name[64];
 #endif
+    int mmWidth, mmHeight;
 
     if (!RRScreenInit (pScreen))
        return FALSE;
@@ -818,6 +820,9 @@ vfbRandRInit(ScreenPtr pScreen)
     pScrPriv->rrOutputValidateMode = vfbRROutputValidateMode;
     pScrPriv->rrModeDestroy = NULL;
 
+    mmWidth = pScreen->width * 25.4 / monitorResolution;
+    mmHeight = pScreen->height * 25.4 / monitorResolution;
+
     RRScreenSetSizeRange (pScreen,
                          1, 1,
                          pScreen->width, pScreen->height);
@@ -850,6 +855,8 @@ vfbRandRInit(ScreenPtr pScreen)
        return FALSE;
     if (!RROutputSetConnection (output, RR_Connected))
        return FALSE;
+    if (!RROutputSetPhysicalSize (output, mmWidth, mmHeight))
+       return FALSE;
     RRCrtcNotify (crtc, mode, 0, 0, RR_Rotate_0, NULL, 1, &output);
 #endif
     return TRUE;
@@ -957,6 +964,9 @@ InitOutput(ScreenInfo * screen_info, int argc, char **argv)
 {
     int i;
     int NumFormats = 0;
+    
+    if (!monitorResolution)
+               monitorResolution = 96;
 
     /* initialize pixmap formats */
 
-- 
GitLab

