From 7d325db5d405e62b4c76e3a6c7dc48bfa9ba1eda Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Fri, 25 Feb 2022 01:18:13 -0500
Subject: [PATCH 1/5] glamor: Remove unused glamor_context::surface

Reviewed-by: Simon Ser <contact@emersion.fr>
---
 glamor/glamor_context.h | 1 -
 1 file changed, 1 deletion(-)

diff --git a/glamor/glamor_context.h b/glamor/glamor_context.h
index 50986971c..45141987f 100644
--- a/glamor/glamor_context.h
+++ b/glamor/glamor_context.h
@@ -38,7 +38,6 @@
 struct glamor_context {
     EGLDisplay display;
     EGLContext ctx;
-    EGLSurface surface;
 
     void (*make_current)(struct glamor_context *glamor_ctx);
 };
-- 
GitLab


From fe831a5e928491e9bd874e26de8891628d035c95 Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Mon, 6 Jun 2022 11:32:18 -0400
Subject: [PATCH 2/5] glamor: Remove unused ::has_pack_invert

Reviewed-by: Simon Ser <contact@emersion.fr>
---
 glamor/glamor.c      | 2 --
 glamor/glamor_priv.h | 1 -
 2 files changed, 3 deletions(-)

diff --git a/glamor/glamor.c b/glamor/glamor.c
index 9dcef5fac..d7473c0c7 100644
--- a/glamor/glamor.c
+++ b/glamor/glamor.c
@@ -760,8 +760,6 @@ glamor_init(ScreenPtr screen, unsigned int flags)
         glamor_priv->has_rw_pbo = TRUE;
 
     glamor_priv->has_khr_debug = epoxy_has_gl_extension("GL_KHR_debug");
-    glamor_priv->has_pack_invert =
-        epoxy_has_gl_extension("GL_MESA_pack_invert");
     glamor_priv->has_fbo_blit =
         epoxy_has_gl_extension("GL_EXT_framebuffer_blit");
     glamor_priv->has_map_buffer_range =
diff --git a/glamor/glamor_priv.h b/glamor/glamor_priv.h
index 028a6d374..9758fcdb5 100644
--- a/glamor/glamor_priv.h
+++ b/glamor/glamor_priv.h
@@ -202,7 +202,6 @@ struct glamor_saved_procs {
 typedef struct glamor_screen_private {
     Bool is_gles;
     int glsl_version;
-    Bool has_pack_invert;
     Bool has_fbo_blit;
     Bool has_map_buffer_range;
     Bool has_buffer_storage;
-- 
GitLab


From c03decc1daf4941c699b647b93c875c0fe67feab Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Mon, 6 Jun 2022 11:54:02 -0400
Subject: [PATCH 3/5] glamor: Remove unused ::has_fbo_blit

Reviewed-by: Simon Ser <contact@emersion.fr>
---
 glamor/glamor.c      | 2 --
 glamor/glamor_priv.h | 1 -
 2 files changed, 3 deletions(-)

diff --git a/glamor/glamor.c b/glamor/glamor.c
index d7473c0c7..6d42788fb 100644
--- a/glamor/glamor.c
+++ b/glamor/glamor.c
@@ -760,8 +760,6 @@ glamor_init(ScreenPtr screen, unsigned int flags)
         glamor_priv->has_rw_pbo = TRUE;
 
     glamor_priv->has_khr_debug = epoxy_has_gl_extension("GL_KHR_debug");
-    glamor_priv->has_fbo_blit =
-        epoxy_has_gl_extension("GL_EXT_framebuffer_blit");
     glamor_priv->has_map_buffer_range =
         epoxy_has_gl_extension("GL_ARB_map_buffer_range") ||
         epoxy_has_gl_extension("GL_EXT_map_buffer_range");
diff --git a/glamor/glamor_priv.h b/glamor/glamor_priv.h
index 9758fcdb5..77e1ca942 100644
--- a/glamor/glamor_priv.h
+++ b/glamor/glamor_priv.h
@@ -202,7 +202,6 @@ struct glamor_saved_procs {
 typedef struct glamor_screen_private {
     Bool is_gles;
     int glsl_version;
-    Bool has_fbo_blit;
     Bool has_map_buffer_range;
     Bool has_buffer_storage;
     Bool has_khr_debug;
-- 
GitLab


From 46dcdac920e1ba66901c972d3fdd94e437efcf99 Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Mon, 6 Jun 2022 12:04:35 -0400
Subject: [PATCH 4/5] glamor: Recognize the GLES spellings of a few extension
 names

Reviewed-by: Simon Ser <contact@emersion.fr>
---
 glamor/glamor.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/glamor/glamor.c b/glamor/glamor.c
index 6d42788fb..15939cd6f 100644
--- a/glamor/glamor.c
+++ b/glamor/glamor.c
@@ -764,7 +764,8 @@ glamor_init(ScreenPtr screen, unsigned int flags)
         epoxy_has_gl_extension("GL_ARB_map_buffer_range") ||
         epoxy_has_gl_extension("GL_EXT_map_buffer_range");
     glamor_priv->has_buffer_storage =
-        epoxy_has_gl_extension("GL_ARB_buffer_storage");
+        epoxy_has_gl_extension("GL_ARB_buffer_storage") ||
+        epoxy_has_gl_extension("GL_EXT_buffer_storage");
     glamor_priv->has_mesa_tile_raster_order =
         epoxy_has_gl_extension("GL_MESA_tile_raster_order");
     glamor_priv->has_nv_texture_barrier =
@@ -782,7 +783,8 @@ glamor_init(ScreenPtr screen, unsigned int flags)
         epoxy_has_gl_extension("GL_ARB_blend_func_extended");
     glamor_priv->has_clear_texture =
         epoxy_gl_version() >= 44 ||
-        epoxy_has_gl_extension("GL_ARB_clear_texture");
+        epoxy_has_gl_extension("GL_ARB_clear_texture") ||
+        epoxy_has_gl_extension("GL_EXT_clear_texture");
 
     glamor_priv->can_copyplane = (gl_version >= 30);
 
-- 
GitLab


From b4748b8906d704c3a43ba8fcdde0b9a6f7cc9419 Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Thu, 20 Jan 2022 17:13:53 -0500
Subject: [PATCH 5/5] xwayland/glamor: Fix our glamor_egl_fd_name_from_pixmap

The return value is a file descriptor, 0 is stdin, don't do that.

Reviewed-by: Simon Ser <contact@emersion.fr>
---
 hw/xwayland/xwayland-glamor.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/hw/xwayland/xwayland-glamor.c b/hw/xwayland/xwayland-glamor.c
index 24620605d..779ff7697 100644
--- a/hw/xwayland/xwayland-glamor.c
+++ b/hw/xwayland/xwayland-glamor.c
@@ -366,7 +366,7 @@ glamor_egl_fd_name_from_pixmap(ScreenPtr screen,
                                PixmapPtr pixmap,
                                CARD16 *stride, CARD32 *size)
 {
-    return 0;
+    return -1;
 }
 
 Bool
-- 
GitLab

