diff -up glibc-2.33/sunrpc/netname.c.omv~ glibc-2.33/sunrpc/netname.c
--- glibc-2.33/sunrpc/netname.c.omv~	2021-05-11 21:34:01.180317491 +0200
+++ glibc-2.33/sunrpc/netname.c	2021-05-11 21:36:17.125314747 +0200
@@ -49,7 +49,12 @@ user2netname (char netname[MAXNETNAMELEN
   if ((strlen (dfltdom) + OPSYS_LEN + 3 + MAXIPRINT) > (size_t) MAXNETNAMELEN)
     return 0;

+  /* The check above makes sure the line below is actually ok, but
+   * gcc 11.1 spews a warning... */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wformat-overflow"
   sprintf (netname, "%s.%d@%s", OPSYS, uid, dfltdom);
+#pragma GCC diagnostic pop
   i = strlen (netname);
   if (netname[i - 1] == '.')
     netname[i - 1] = '\0';
