--- glibc-2.35/malloc/malloc.c~	2022-02-03 05:27:54.000000000 +0000
+++ glibc-2.35/malloc/malloc.c	2022-03-09 15:04:18.825885223 +0000
@@ -2586,12 +2586,12 @@
 	{
 	  /* There is no need to isse the THP madvise call if Huge Pages are
 	     used directly.  */
-	  mm = sysmalloc_mmap (nb, mp_.hp_pagesize, mp_.hp_flags, av);
+	  mm = sysmalloc_mmap (nb, mp_.hp_pagesize, mp_.hp_flags | MAP_POPULATE, av);
 	  if (mm != MAP_FAILED)
 	    return mm;
 	}
 #endif
-      mm = sysmalloc_mmap (nb, pagesize, 0, av);
+      mm = sysmalloc_mmap (nb, pagesize, MAP_POPULATE, av);
       if (mm != MAP_FAILED)
 	return mm;
       tried_mmap = true;
--- glibc-2.37/sysdeps/generic/adaptive_spin_count.h~	2023-02-01 03:27:45.000000000 +0000
+++ glibc-2.37/sysdeps/generic/adaptive_spin_count.h	2023-04-20 16:24:07.132939021 +0000
@@ -19,4 +19,4 @@
 /* The choice of 100 spins for the default spin count for an adaptive spin
    is a completely arbitrary choice that has not been evaluated thoroughly
    using modern hardware.  */
-#define DEFAULT_ADAPTIVE_COUNT 100
+#define DEFAULT_ADAPTIVE_COUNT 5000
