--- meson.options.orig	2025-07-04 23:52:54.275168711 +0200
+++ meson.options	2025-07-04 23:51:49.809054691 +0200
@@ -39,6 +39,11 @@ option(
   description : 'Use X86 SSSE3 intrinsic optimized paths',
 )
 option(
+  'avx2',
+  type : 'feature',
+  description : 'Use X86 AVX2 intrinsic optimized paths',
+)
+option(
   'vmx',
   type : 'feature',
   description : 'Use PPC VMX/Altivec intrinsic optimized paths',
