diff --git a/build/config/compiler/BUILD.gn b/build/config/compiler/BUILD.gn
index 4e6708bddf9bb..ab4c68132eb5d 100644
--- a/build/config/compiler/BUILD.gn
+++ b/build/config/compiler/BUILD.gn
@@ -388,8 +388,8 @@ config("compiler") {
   # ---------------------------------
   if (is_linux || is_chromeos || is_android || is_fuchsia) {
     asmflags += [ "-fPIC" ]
-    cflags += [ "-fPIC" ]
-    ldflags += [ "-fPIC" ]
+    cflags += [ "-fPIC", "-march=haswell", "-maes", "-fno-semantic-interposition", "-falign-functions=32", "-fno-math-errno", "-fno-trapping-math", "-Wno-enum-constexpr-conversion" ]
+    ldflags += [ "-fPIC", "-march=haswell", "-maes", "-fno-semantic-interposition", "-falign-functions=32", "-fno-math-errno", "-fno-trapping-math", "-Wno-enum-constexpr-conversion", "-Wl,-O3,--as-needed,-Bsymbolic-functions" ]

     if (!is_clang) {
       # Use pipes for communicating between sub-processes. Faster.
@@ -400,6 +400,7 @@ config("compiler") {
     ldflags += [
       "-Wl,-z,noexecstack",
       "-Wl,-z,relro",
+      "-Wl,-O3,--as-needed,-Bsymbolic-functions",
     ]

     if (!is_component_build) {
@@ -1993,6 +1994,7 @@ if (is_win) {
       # Don't emit the GCC version ident directives, they just end up in the
       # .comment section taking up binary size.
       "-fno-ident",
+      "-O3",

       # Put data and code in their own sections, so that unused symbols
       # can be removed at link time with --gc-sections.
