diff --git a/build/config/rust.gni b/build/config/rust.gni
index 3b5db82..784ea1c 100644
--- a/build/config/rust.gni
+++ b/build/config/rust.gni
@@ -63,6 +63,11 @@ declare_args() {
   # the bindgen exectuable).
   rust_bindgen_root = "//third_party/rust-toolchain"
 
+  # Optional directory containing libclang.so. This is separate from
+  # rust_bindgen_root so distributions can use a system bindgen executable with
+  # the libclang that exactly matches clang_base_path.
+  bindgen_libclang_path = ""
+
   # If you're using a Rust toolchain as specified by rust_sysroot_absolute,
   # set this to the output of `rustc -V`. Changing this string will cause all
   # Rust targets to be rebuilt, which allows you to update your toolchain and
diff --git a/build/rust/rust_bindgen_generator.gni b/build/rust/rust_bindgen_generator.gni
index 5716763..cf6d387 100644
--- a/build/rust/rust_bindgen_generator.gni
+++ b/build/rust/rust_bindgen_generator.gni
@@ -35,6 +35,9 @@ if (!use_chromium_rust_toolchain &&
 } else {
   _libclang_path += "/lib"
 }
+if (bindgen_libclang_path != "") {
+  _libclang_path = bindgen_libclang_path
+}
 
 # Template to build Rust/C bindings with bindgen.
 #
