diff --git a/Makefile b/Makefile
--- a/Makefile
+++ b/Makefile
@@ -449,9 +449,9 @@ HOSTRUSTC = rustc
 HOSTPKG_CONFIG	= pkg-config

 KBUILD_USERHOSTCFLAGS := -Wall -Wmissing-prototypes -Wstrict-prototypes \
-			 -O2 -fomit-frame-pointer -std=gnu11
+			 -O3 -march=native -mtune=native -mllvm -inline-threshold=1000 -maes -mbmi2 -mpclmul -mllvm -extra-vectorizer-passes -mllvm -enable-cond-stores-vec -mllvm -slp-vectorize-hor-store -mllvm -enable-loopinterchange -mllvm -enable-loop-distribute -mllvm -enable-unroll-and-jam -mllvm -enable-loop-flatten -mllvm -unroll-runtime-multi-exit -mllvm -aggressive-ext-opt -mllvm -enable-interleaved-mem-accesses -mllvm -enable-masked-interleaved-mem-accesses -fno-math-errno -fno-trapping-math -falign-functions=32 -fno-semantic-interposition -fcf-protection=none -mharden-sls=none -fomit-frame-pointer -mllvm -adce-remove-loops -mllvm -enable-ext-tsp-block-placement -mllvm -enable-gvn-hoist -mllvm -enable-dfa-jump-thread -mprefer-vector-width=256 -std=gnu11 -fprofile-use=/home/marcus/Downloads/all.profdata -Wno-error -ffp-contract=fast -fdata-sections -ffunction-sections -fsplit-machine-functions -mtls-dialect=gnu2 -w
 KBUILD_USERCFLAGS  := $(KBUILD_USERHOSTCFLAGS) $(USERCFLAGS)
-KBUILD_USERLDFLAGS := $(USERLDFLAGS)
+KBUILD_USERLDFLAGS := -Wl,--lto-O3,-O3,-Bsymbolic-functions,--as-needed -Wl,--lto-CGO3 -Wl,--gc-sections -Wl,--icf=all -Wl,-z,now -Wl,-z,relro -Wl,-z,pack-relative-relocs -Wl,--hash-style=gnu -fuse-ld=lld -fprofile-use=/home/marcus/Downloads/all.profdata -w $(USERLDFLAGS)

 # These flags apply to all Rust code in the tree, including the kernel and
 # host programs.
@@ -460,6 +460,7 @@ export rust_common_flags := --edition=20
 			    -Dunsafe_op_in_unsafe_fn -Drust_2018_idioms \
 			    -Dunreachable_pub -Dnon_ascii_idents \
 			    -Wmissing_docs \
+			    -Ctarget-cpu=native \
 			    -Drustdoc::missing_crate_level_docs \
 			    -Dclippy::correctness -Dclippy::style \
 			    -Dclippy::suspicious -Dclippy::complexity \
@@ -471,11 +472,11 @@ export rust_common_flags := --edition=20
 			    -Wclippy::dbg_macro

 KBUILD_HOSTCFLAGS   := $(KBUILD_USERHOSTCFLAGS) $(HOST_LFS_CFLAGS) $(HOSTCFLAGS)
-KBUILD_HOSTCXXFLAGS := -Wall -O2 $(HOST_LFS_CFLAGS) $(HOSTCXXFLAGS)
+KBUILD_HOSTCXXFLAGS := -O3 -march=native -mtune=native -mllvm -inline-threshold=1000 -maes -mbmi2 -mpclmul -mllvm -extra-vectorizer-passes -mllvm -enable-cond-stores-vec -mllvm -slp-vectorize-hor-store -mllvm -enable-loopinterchange -mllvm -enable-loop-distribute -mllvm -enable-unroll-and-jam -mllvm -enable-loop-flatten -mllvm -unroll-runtime-multi-exit -mllvm -aggressive-ext-opt -mllvm -enable-interleaved-mem-accesses -mllvm -enable-masked-interleaved-mem-accesses -fno-math-errno -fno-trapping-math -falign-functions=32 -fno-semantic-interposition -fcf-protection=none -mharden-sls=none -fomit-frame-pointer -mllvm -adce-remove-loops -mllvm -enable-ext-tsp-block-placement -mllvm -enable-gvn-hoist -mllvm -enable-dfa-jump-thread -mprefer-vector-width=256 -fprofile-use=/home/marcus/Downloads/all.profdata -Wno-error -ffp-contract=fast -fdata-sections -ffunction-sections -fsplit-machine-functions -mtls-dialect=gnu2 -w $(HOST_LFS_CFLAGS) $(HOSTCXXFLAGS)
 KBUILD_HOSTRUSTFLAGS := $(rust_common_flags) -O -Cstrip=debuginfo \
 			-Zallow-features= $(HOSTRUSTFLAGS)
-KBUILD_HOSTLDFLAGS  := $(HOST_LFS_LDFLAGS) $(HOSTLDFLAGS)
-KBUILD_HOSTLDLIBS   := $(HOST_LFS_LIBS) $(HOSTLDLIBS)
+KBUILD_HOSTLDFLAGS  := -Wl,--lto-O3,-O3,-Bsymbolic-functions,--as-needed -Wl,--lto-CGO3 -Wl,--gc-sections -Wl,--icf=all -Wl,-z,now -Wl,-z,relro -Wl,-z,pack-relative-relocs -Wl,--hash-style=gnu -fuse-ld=lld -fprofile-use=/home/marcus/Downloads/all.profdata -w $(HOST_LFS_LDFLAGS) $(HOSTLDFLAGS)
+KBUILD_HOSTLDLIBS   := -Wl,--lto-O3,-O3,-Bsymbolic-functions,--as-needed -Wl,--lto-CGO3 -Wl,--gc-sections -Wl,--icf=all -Wl,-z,now -Wl,-z,relro -Wl,-z,pack-relative-relocs -Wl,--hash-style=gnu -fuse-ld=lld -fprofile-use=/home/marcus/Downloads/all.profdata -w $(HOST_LFS_LIBS) $(HOSTLDLIBS)

 # Make variables (CC, etc...)
 CPP		= $(CC) -E
@@ -527,14 +528,14 @@ PAHOLE_FLAGS	= $(shell PAHOLE=$(PAHOLE)
 CHECKFLAGS     := -D__linux__ -Dlinux -D__STDC__ -Dunix -D__unix__ \
 		  -Wbitwise -Wno-return-void -Wno-unknown-attribute $(CF)
 NOSTDINC_FLAGS :=
-CFLAGS_MODULE   =
-RUSTFLAGS_MODULE =
-AFLAGS_MODULE   =
-LDFLAGS_MODULE  =
-CFLAGS_KERNEL	=
-RUSTFLAGS_KERNEL =
-AFLAGS_KERNEL	=
-LDFLAGS_vmlinux =
+CFLAGS_MODULE   = -O3 -march=native -mtune=native -mllvm -inline-threshold=1000 -mllvm -extra-vectorizer-passes -mllvm -enable-cond-stores-vec -mllvm -slp-vectorize-hor-store -mllvm -enable-loopinterchange -mllvm -enable-loop-distribute -mllvm -enable-unroll-and-jam -mllvm -enable-loop-flatten -mllvm -unroll-runtime-multi-exit -mllvm -aggressive-ext-opt -mllvm -enable-interleaved-mem-accesses -mllvm -enable-masked-interleaved-mem-accesses -fno-math-errno -fno-trapping-math -falign-functions=32 -fno-semantic-interposition -fcf-protection=none -mharden-sls=none -fomit-frame-pointer -mprefer-vector-width=256 -mllvm -adce-remove-loops -mllvm -enable-ext-tsp-block-placement -mllvm -enable-gvn-hoist -mllvm -enable-dfa-jump-thread -std=gnu11 -fprofile-use=/home/marcus/Downloads/all.profdata -Wno-error -ffp-contract=fast -fdata-sections -ffunction-sections -fsplit-machine-functions -mtls-dialect=gnu2 -w
+RUSTFLAGS_MODULE = -Ctarget-cpu=native
+AFLAGS_MODULE   = -D__AVX__=1 -D__AVX2__=1 -D__FMA__=1
+LDFLAGS_MODULE  =
+CFLAGS_KERNEL	= -march=native -mtune=native -mllvm -inline-threshold=1000 -mllvm -extra-vectorizer-passes -mllvm -enable-cond-stores-vec -mllvm -slp-vectorize-hor-store -mllvm -enable-loopinterchange -mllvm -enable-loop-distribute -mllvm -enable-unroll-and-jam -mllvm -enable-loop-flatten -mllvm -unroll-runtime-multi-exit -mllvm -aggressive-ext-opt -mllvm -enable-interleaved-mem-accesses -mllvm -enable-masked-interleaved-mem-accesses -fno-math-errno -fno-trapping-math -falign-functions=32 -fno-semantic-interposition -fcf-protection=none -mharden-sls=none -fomit-frame-pointer -mprefer-vector-width=256 -mllvm -adce-remove-loops -mllvm -enable-ext-tsp-block-placement -mllvm -enable-gvn-hoist -mllvm -enable-dfa-jump-thread -std=gnu11 -fprofile-use=/home/marcus/Downloads/all.profdata -Wno-error -ffp-contract=fast -fdata-sections -ffunction-sections -fsplit-machine-functions -mtls-dialect=gnu2 -w
+RUSTFLAGS_KERNEL = -Ctarget-cpu=native
+AFLAGS_KERNEL	= -D__AVX__=1 -D__AVX2__=1 -D__FMA__=1
+LDFLAGS_vmlinux = -Wl,--lto-O3,-O3,-Bsymbolic-functions,--as-needed -Wl,--lto-CGO3 -Wl,--gc-sections -Wl,--icf=all -Wl,-z,now -Wl,-z,relro -Wl,-z,pack-relative-relocs -Wl,--hash-style=gnu -march=native -fcf-protection=none -mharden-sls=none -fuse-ld=lld -fprofile-use=/home/marcus/Downloads/all.profdata -w

 # Use USERINCLUDE when you must reference the UAPI directories only.
 USERINCLUDE    := \
@@ -569,6 +570,7 @@ KBUILD_RUSTFLAGS := $(rust_common_flags)
 		    --target=$(objtree)/scripts/target.json \
 		    -Cpanic=abort -Cembed-bitcode=n -Clto=n \
 		    -Cforce-unwind-tables=n -Ccodegen-units=1 \
+		    -Ctarget-cpu=native \
 		    -Csymbol-mangling-version=v0 \
 		    -Crelocation-model=static \
 		    -Zfunction-sections=n \
