diff --git a/kernel/sched/alt_core.c b/kernel/sched/alt_core.c
index 7b2ac7496..8361027a1 100644
--- a/kernel/sched/alt_core.c
+++ b/kernel/sched/alt_core.c
@@ -1943,10 +1943,13 @@ static inline int select_task_rq(struct task_struct *p)
 	if (unlikely(!cpumask_and(&allow_mask, p->cpus_ptr, cpu_active_mask)))
 		return select_fallback_rq(task_cpu(p), p);
 
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wuninitialized"
 	if ((cpumask_intersects(&mask, sched_idle_mask) &&
 	     static_call(sched_idle_select_func)(&mask, &allow_mask, sched_idle_mask)) ||
 	    preempt_mask_check(&mask, &allow_mask, task_sched_prio(p)))
 		return best_mask_cpu(task_cpu(p), &mask);
+#pragma clang diagnostic pop
 
 	return best_mask_cpu(task_cpu(p), &allow_mask);
 }
