--- a/Makefile.in	2026-05-10 16:19:19.178649179 +0200
+++ b/Makefile.in	2026-05-10 16:41:04.969382996 +0200
@@ -51,11 +51,11 @@ DIST_VERSION := $(DST_BASE)/version
 DEPLOY_DIR := ./deploy
 REDIST_DIR := ./redist
 
-HOST_CFLAGS    ?= -O2 -march=nocona -mtune=core-avx2
-CCOS_CFLAGS    := $(HOST_CFLAGS) -pipe -mno-avx512f
+HOST_CFLAGS    ?= -O3 -march=native -mtune=native
+CCOS_CFLAGS    := $(HOST_CFLAGS)
 i386_CFLAGS    := -mstackrealign $(CCOS_CFLAGS) -mfpmath=sse
 x86_64_CFLAGS  := -mcmodel=small $(CCOS_CFLAGS) -mfpmath=sse
-HOST_RUSTFLAGS   ?= -Copt-level=3 -Ctarget-cpu=nocona
+HOST_RUSTFLAGS   ?= -Copt-level=3 -Ctarget-cpu=native
 i386_RUSTFLAGS   := $(HOST_RUSTFLAGS)
 x86_64_RUSTFLAGS := $(HOST_RUSTFLAGS)
 aarch64_CFLAGS := -march=armv8.2-a -mtune=cortex-x3
@@ -64,7 +64,7 @@ aarch64_RUSTFLAGS := -Copt-level=3 -Ctar
 arm64ec_RUSTFLAGS := -Copt-level=3 -Ctarget-cpu=armv8.2-a
 
 CFLAGS  = -fwrapv -fno-strict-aliasing
-CFLAGS += -ffunction-sections -fdata-sections -fno-omit-frame-pointer
+CFLAGS += -ffunction-sections -fdata-sections
 CFLAGS += -ffile-prefix-map=$(CCACHE_BASEDIR)=.
 
 ifneq ($(SUPPRESS_WARNINGS),)
@@ -116,9 +116,9 @@ ifeq ($(shell cc --version|grep 'clang')
 	TOOLCHAIN_PREFIX :=
 	OBJCOPY_FLAGS  := -p
 	MINGW_TYPE := gcc
-	i386_CFLAGS += -s -mno-avx -mno-avx2
+	i386_CFLAGS += -s
 	i386_CFLAGS += -fvect-cost-model=cheap -fno-semantic-interposition -fipa-pta
-	x86_64_CFLAGS += -s -mno-avx -mno-avx2
+	x86_64_CFLAGS += -s
 	x86_64_CFLAGS += -fvect-cost-model=cheap -fno-semantic-interposition -fipa-pta
 	i386_SANITY_FLAGS := -mpreferred-stack-boundary=2
 else
@@ -827,6 +827,16 @@ $(eval $(call rules-cmake,spirv-headers,
 ## glslang
 ##
 
+GLSLANG_CMAKE_ARGS = \
+	-DCMAKE_BUILD_TYPE=Release \
+	-DCMAKE_C_COMPILER=clang \
+	-DCMAKE_CXX_COMPILER=clang++ \
+	-DCMAKE_C_FLAGS_RELEASE='-DNDEBUG -O3 -march=native -mtune=native -std=gnu17 -fno-semantic-interposition -ffunction-sections -fdata-sections -flto -fprofile-use=/home/marcus/Downloads/glslang.profdata' \
+	-DCMAKE_CXX_FLAGS_RELEASE='-DNDEBUG -O3 -march=native -mtune=native -std=gnu++20 -fno-semantic-interposition -ffunction-sections -fdata-sections -flto -fprofile-use=/home/marcus/Downloads/glslang.profdata' \
+	-DCMAKE_EXE_LINKER_FLAGS_RELEASE='-fuse-ld=lld -Wl,-O2,--as-needed,--gc-sections,-z,max-page-size=0x200000 -fprofile-use=/home/marcus/Downloads/glslang.profdata' \
+	-DCMAKE_SHARED_LINKER_FLAGS_RELEASE='-fuse-ld=lld -Wl,-O2,--as-needed,--gc-sections,-z,max-page-size=0x200000 -fprofile-use=/home/marcus/Downloads/glslang.profdata' \
+	-DCMAKE_MODULE_LINKER_FLAGS_RELEASE='-fuse-ld=lld -Wl,-O2,--as-needed,--gc-sections,-z,max-page-size=0x200000 -fprofile-use=/home/marcus/Downloads/glslang.profdata'
+
 $(eval $(call rules-source,glslang,$(SRCDIR)/glslang))
 $(eval $(call rules-cmake,glslang,i386,unix))
 $(eval $(call rules-cmake,glslang,x86_64,unix))
@@ -1314,7 +1324,7 @@ D7VK_MESON_ARGS += -Denable_dxgi=false -
 D7VK_i386_MESON_ARGS = --bindir=$(D7VK_i386_DST)/lib/wine/d7vk/i386-windows
 D7VK_HOST_DEPENDS = glslang
 
-D7VK_i386_CFLAGS = -O3 -mno-avx $(i386_SANITY_FLAGS)
+D7VK_i386_CFLAGS = -O3 $(i386_SANITY_FLAGS)
 D7VK_CPPFLAGS = -msse -msse2
 D7VK_LDFLAGS = -static -static-libgcc -static-libstdc++
 
