--- a/make/rules-cmake.mk	2023-10-12 16:17:36.260323696 +0200
+++ b/make/rules-cmake.mk	2023-10-12 16:17:49.534329154 +0200
@@ -16,7 +16,7 @@
 	cmake "$$($(2)_SRC)" \
 	    -DCMAKE_INSTALL_PREFIX="$$($(2)_DST$(3))" \
 	    -DCMAKE_INSTALL_LIBDIR="lib$(subst 32,,$(3))" \
-	    -DCMAKE_BUILD_TYPE=plain \
+	    -DCMAKE_BUILD_TYPE=release \
 	    -DCMAKE_SYSTEM_NAME=$(if $(4),Windows,) \
 	    -DCMAKE_SHARED_LIBRARY_PREFIX_C=$(if $(4),,lib) \
 	    -DCMAKE_IMPORT_LIBRARY_PREFIX_C=$(if $(4),,lib) \


--- a/make/rules-meson.mk	2023-10-11 15:59:05.000000000 +0200
+++ b/make/rules-meson.mk	2023-10-12 16:27:43.666517375 +0200
@@ -50,7 +50,7 @@
 	meson "$$($(2)_OBJ$(3))" "$$($(2)_SRC)" \
 	      --prefix="$$($(2)_DST$(3))" \
 	      --libdir="lib$(subst 32,,$(3))" \
-	      --buildtype=plain \
+	      --buildtype=release \
 	      $(if $(4),--cross-file="$$($(2)_OBJ$(3))/cross-$(3).txt",) \
 	      $$($(2)_MESON_ARGS) \
 	      $$($(2)_MESON_ARGS$(3)) \

diff --git a/Makefile.in b/Makefile.in
index faee5a57..4e5c34e5 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -313,7 +313,7 @@ $(eval $(call rules-cmake,spirv-headers,
 ## Vulkan-Loader
 ##
 
-VULKAN_LOADER_CMAKE_ARGS = -DUSE_MASM=OFF
+VULKAN_LOADER_CMAKE_ARGS = -DUSE_MASM=OFF -DCMAKE_BUILD_TYPE=Release -DJWASM_FOUND=0 -DCMAKE_C_FLAGS_RELEASE='-DNDEBUG -flto=auto -fivopts -fmodulo-sched -std=gnu18'
 VULKAN_LOADER_CMAKE_ARGS64 = -DVULKAN_HEADERS_INSTALL_DIR=$(VULKAN_HEADERS_DST64)
 VULKAN_LOADER_CMAKE_ARGS32 = -DVULKAN_HEADERS_INSTALL_DIR=$(VULKAN_HEADERS_DST32)
 VULKAN_LOADER_CFLAGS = -DWINVER=0x0A00 -D_WIN32_WINNT=0x0A00 # 0x0A00 is _WIN32_WINNT_WIN10
@@ -328,6 +328,8 @@ $(eval $(call rules-cmake,vulkan-loader,
 ## glslang
 ##
 
+GLSLANG_CMAKE_ARGS = -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_FLAGS_RELEASE='-DNDEBUG -flto=auto -fivopts -fmodulo-sched -Wl,-zmax-page-size=0x200000 -std=gnu18' -DCMAKE_CXX_FLAGS_RELEASE='-DNDEBUG -flto=auto -fivopts -fmodulo-sched -Wl,-zmax-page-size=0x200000 -std=gnu++2a'
+
 $(eval $(call rules-source,glslang,$(SRCDIR)/glslang))
 $(eval $(call rules-cmake,glslang,32))
 $(eval $(call rules-cmake,glslang,64))
@@ -338,6 +340,9 @@ $(eval $(call rules-cmake,glslang,64))
 ##
 
 LSTEAMCLIENT_DEPENDS = wine
+LSTEAMCLIENT_CFLAGS = -fivopts -fmodulo-sched -std=gnu18
+LSTEAMCLIENT_CXXFLAGS = -fivopts -fmodulo-sched -std=gnu++2a
+LSTEAMCLIENT_LDFLAGS = -static-libgcc -static-libstdc++ -ldl
 
 $(eval $(call rules-source,lsteamclient,$(SRCDIR)/lsteamclient))
 $(eval $(call rules-makedep,lsteamclient,32))
@@ -389,6 +394,8 @@ STEAMEXE_LDFLAGS = \
 	-L$(SRC)/steam_helper/64/ \
 	-lsteam_api
 
+STEAMEXE_CFLAGS = -fivopts -fmodulo-sched -std=gnu18
+STEAMEXE_CXXFLAGS = -fivopts -fmodulo-sched -std=gnu++2a
 STEAMEXE_DEPENDS = wine
 
 $(eval $(call rules-source,steamexe,$(SRCDIR)/steam_helper))
@@ -441,6 +448,9 @@ WINE_CONFIGURE_ARGS64 = --enable-win64 \
 
 WINE_DEPENDS = gst_orc gstreamer gst_base vkd3d
 
+WINE_CFLAGS = -fivopts -fmodulo-sched -std=gnu18
+WINE_CXXFLAGS = -fivopts -fmodulo-sched -std=gnu++2a
+
 $(eval $(call rules-source,wine,$(SRCDIR)/wine))
 $(eval $(call rules-autoconf,wine,32))
 $(eval $(call rules-autoconf,wine,64))
@@ -471,6 +481,8 @@ $(OBJ)/.wine-post-build32:
 ##
 
 VRCLIENT_DEPENDS = wine
+VRCLIENT_CFLAGS = -fivopts -fmodulo-sched -std=gnu18
+VRCLIENT_CXXFLAGS = -fivopts -fmodulo-sched -std=gnu++2a
 
 $(eval $(call rules-source,vrclient,$(SRCDIR)/vrclient_x64))
 $(eval $(call rules-makedep,vrclient,32))
@@ -557,7 +569,9 @@ VKD3D_CONFIGURE_ARGS = \
   --disable-demos \
   --without-ncurses \
 
-VKD3D_LDFLAGS = -static-libgcc $(CROSSLDFLAGS)
+VKD3D_CFLAGS = -flto=auto -fivopts -fmodulo-sched -std=gnu18
+VKD3D_CXXFLAGS = -flto=auto -fivopts -fmodulo-sched -std=gnu++2a
+VKD3D_LDFLAGS = -static-libgcc -flto=auto $(CROSSLDFLAGS)
 VKD3D_DEPENDS = vulkan-loader vulkan-headers spirv-headers
 
 $(eval $(call rules-source,vkd3d,$(SRCDIR)/vkd3d))
@@ -593,7 +607,9 @@ ifneq ($(UNSTRIPPED_BUILD),)
 endif
 
 VKD3D_PROTON_CPPFLAGS = -msse -msse2
-VKD3D_PROTON_LDFLAGS = -static -static-libgcc -static-libstdc++
+VKD3D_PROTON_CFLAGS = -flto=auto -fivopts -fmodulo-sched -std=gnu18
+VKD3D_PROTON_CXXFLAGS = -flto=auto -fivopts -fmodulo-sched -std=gnu++2a
+VKD3D_PROTON_LDFLAGS = -static -static-libgcc -static-libstdc++ -flto=auto
 
 $(eval $(call rules-source,vkd3d-proton,$(SRCDIR)/vkd3d-proton))
 $(eval $(call rules-meson,vkd3d-proton,32,CROSS))
