diff --git a/Makefile.in b/Makefile.in
index faee5a57..4e5c34e5 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -448,7 +448,7 @@ $(eval $(call rules-cmake,spirv-headers,
 ## Vulkan-Loader
 ##

-VULKAN_LOADER_CMAKE_ARGS = -DUSE_MASM=OFF
+VULKAN_LOADER_CMAKE_ARGS = -DUSE_MASM=OFF -DCMAKE_BUILD_TYPE=Release -DJWASM_FOUND=0 -DCMAKE_C_FLAGS_RELEASE='-DNDEBUG -flto=auto -fivopts -fmodulo-sched'
 VULKAN_LOADER_CMAKE_ARGS64 = -DVULKAN_HEADERS_INSTALL_DIR=$(VULKAN_HEADERS_DST64)
 VULKAN_LOADER_CMAKE_ARGS32 = -DVULKAN_HEADERS_INSTALL_DIR=$(VULKAN_HEADERS_DST32)
 VULKAN_LOADER_CFLAGS = -DWINVER=0x0A00 -D_WIN32_WINNT=0x0A00 # 0x0A00 is _WIN32_WINNT_WIN10
@@ -463,6 +463,7 @@ $(eval $(call rules-cmake,vulkan-loader,
 ## glslang
 ##

+GLSLANG_CMAKE_ARGS = -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_FLAGS_RELEASE='-DNDEBUG -flto=auto -fivopts -fmodulo-sched' -DCMAKE_CXX_FLAGS_RELEASE='-DNDEBUG -flto=auto -fivopts -fmodulo-sched' -DCMAKE_C_STANDARD=23 -DCMAKE_CXX_STANDARD=23
 $(eval $(call rules-source,glslang,$(SRCDIR)/glslang))
 $(eval $(call rules-cmake,glslang,32))
 $(eval $(call rules-cmake,glslang,64))
@@ -472,8 +473,8 @@ $(eval $(call rules-cmake,glslang,64))
 ## lsteamclient
 ##

-LSTEAMCLIENT_CFLAGS = -Wno-attributes
-LSTEAMCLIENT_CXXFLAGS = -Wno-attributes
+LSTEAMCLIENT_CFLAGS = -Wno-attributes -fivopts -fmodulo-sched
+LSTEAMCLIENT_CXXFLAGS = -Wno-attributes -fivopts -fmodulo-sched
 LSTEAMCLIENT_LDFLAGS = -static-libgcc -static-libstdc++ -ldl

 LSTEAMCLIENT_WINEMAKER_ARGS = \
@@ -528,8 +529,8 @@ all-dist: $(DIST_WINEOPENXR_JSON64)
 ## steam.exe
 ##

-STEAMEXE_CFLAGS = -Wno-attributes
-STEAMEXE_CXXFLAGS = -Wno-attributes
+STEAMEXE_CFLAGS = -Wno-attributes -fivopts -fmodulo-sched
+STEAMEXE_CXXFLAGS = -Wno-attributes -fivopts -fmodulo-sched
 STEAMEXE_LDFLAGS = -L$(STEAMEXE_SRC)/32/ -L$(STEAMEXE_SRC)/64/ -lsteam_api -lshlwapi -lmsi -lole32 -ldl -static-libgcc -static-libstdc++

 STEAMEXE_WINEMAKER_ARGS = \
@@ -669,6 +655,9 @@ WINE_CONFIGURE_ARGS64 = --enable-win64 \

 WINE_DEPENDS = gst_orc gstreamer gst_base vkd3d openfst kaldi vosk

+WINE_CFLAGS = -fivopts -fmodulo-sched
+WINE_CXXFLAGS = -fivopts -fmodulo-sched
+
 $(eval $(call rules-source,wine,$(SRCDIR)/wine))
 $(eval $(call rules-autoconf,wine,32))
 $(eval $(call rules-autoconf,wine,64))
@@ -698,8 +687,8 @@ $(OBJ)/.wine-post-build32:
 ## vrclient
 ##

-VRCLIENT_CFLAGS = -Wno-attributes
-VRCLIENT_CXXFLAGS = -Wno-attributes
+VRCLIENT_CFLAGS = -Wno-attributes -fivopts -fmodulo-sched
+VRCLIENT_CXXFLAGS = -Wno-attributes -fivopts -fmodulo-sched
 VRCLIENT_LDFLAGS = -static-libgcc -static-libstdc++ -ldl

 VRCLIENT_WINEMAKER_ARGS = \
@@ -798,7 +774,9 @@ VKD3D_CONFIGURE_ARGS = \
   --disable-demos \
   --without-ncurses \

-VKD3D_LDFLAGS = -static-libgcc $(CROSSLDFLAGS)
+VKD3D_CFLAGS = $(CFLAGS) -flto=auto -fivopts -fmodulo-sched
+VKD3D_CXXFLAGS = $(CXXFLAGS) -flto=auto -fivopts -fmodulo-sched
+VKD3D_LDFLAGS = -static-libgcc -flto=auto $(CROSSLDFLAGS)
 VKD3D_DEPENDS = vulkan-loader vulkan-headers spirv-headers

 $(eval $(call rules-source,vkd3d,$(SRCDIR)/vkd3d))
@@ -834,6 +812,8 @@ ifneq ($(UNSTRIPPED_BUILD),)
 endif

 VKD3D_PROTON_CPPFLAGS = -msse -msse2
+VKD3D_PROTON_CFLAGS = -flto=auto -fivopts -fmodulo-sched
+VKD3D_PROTON_CXXFLAGS = -flto=auto -fivopts -fmodulo-sched
-VKD3D_PROTON_LDFLAGS = -static -static-libgcc -static-libstdc++
+VKD3D_PROTON_LDFLAGS = -static -static-libgcc -static-libstdc++ -flto=auto

 $(eval $(call rules-source,vkd3d-proton,$(SRCDIR)/vkd3d-proton))
