diff --git a/Makefile.in b/Makefile.in
index 6587982f..e87bb4eb 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -448,7 +448,7 @@ $(eval $(call rules-cmake,spirv-headers,
 ## Vulkan-Loader
 ##

-VULKAN_LOADER_CMAKE_ARGS = -DUSE_MASM=OFF
+VULKAN_LOADER_CMAKE_ARGS = -DUSE_MASM=OFF -DCMAKE_BUILD_TYPE=Release -DJWASM_FOUND=0 -DCMAKE_C_FLAGS_RELEASE='-DNDEBUG -flto=auto'
 VULKAN_LOADER_CMAKE_ARGS64 = -DVULKAN_HEADERS_INSTALL_DIR=$(VULKAN_HEADERS_DST64)
 VULKAN_LOADER_CMAKE_ARGS32 = -DVULKAN_HEADERS_INSTALL_DIR=$(VULKAN_HEADERS_DST32)
 VULKAN_LOADER_CFLAGS = -DWINVER=0x0A00 -D_WIN32_WINNT=0x0A00 # 0x0A00 is _WIN32_WINNT_WIN10
@@ -798,7 +798,7 @@ VKD3D_CONFIGURE_ARGS = \
   --disable-demos \
   --without-ncurses \

-VKD3D_LDFLAGS = -static-libgcc $(CROSSLDFLAGS)
+VKD3D_LDFLAGS = -static-libgcc $(CROSSLDFLAGS)
 VKD3D_DEPENDS = vulkan-loader vulkan-headers spirv-headers

 $(eval $(call rules-source,vkd3d,$(SRCDIR)/vkd3d))
