diff --git a/Makefile.in b/Makefile.in
index faee5a57..4e5c34e5 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -448,7 +448,7 @@ $(eval $(call rules-cmake,spirv-headers,
 ## Vulkan-Loader
 ##
 
-VULKAN_LOADER_CMAKE_ARGS = -DUSE_MASM=OFF
+VULKAN_LOADER_CMAKE_ARGS = -DUSE_MASM=OFF -DCMAKE_BUILD_TYPE=Release -DJWASM_FOUND=0 -DCMAKE_C_FLAGS_RELEASE='-DNDEBUG -flto=auto'
 VULKAN_LOADER_CMAKE_ARGS64 = -DVULKAN_HEADERS_INSTALL_DIR=$(VULKAN_HEADERS_DST64)
 VULKAN_LOADER_CMAKE_ARGS32 = -DVULKAN_HEADERS_INSTALL_DIR=$(VULKAN_HEADERS_DST32)
 VULKAN_LOADER_CFLAGS = -DWINVER=0x0A00 -D_WIN32_WINNT=0x0A00 # 0x0A00 is _WIN32_WINNT_WIN10
@@ -636,21 +636,6 @@ $(eval $(call rules-cmake,vosk,64))
 ## wine
 ##
 
-WINE_SOURCE_ARGS = \
-  --exclude configure \
-  --exclude autom4te.cache \
-  --exclude include/config.h.in \
-  --exclude include/config.h.ink \
-  --exclude include/wine/vulkan.h \
-  --exclude include/wine/vulkan_driver.h \
-  --exclude dlls/vulkan-1/vulkan-1.spec \
-  --exclude dlls/winevulkan/loader_thunks.c \
-  --exclude dlls/winevulkan/loader_thunks.h \
-  --exclude dlls/winevulkan/vulkan_thunks.c \
-  --exclude dlls/winevulkan/vulkan_thunks.h \
-  --exclude dlls/winevulkan/winevulkan.json \
-  --exclude dlls/winevulkan/winevulkan.spec
-
 WINE_CONFIGURE_ARGS = \
   --with-x \
   --with-gstreamer \
@@ -778,19 +763,6 @@ $(OBJ)/.dxvk-nvapi-post-build32:
 ## vkd3d
 ##
 
-VKD3D_SOURCE_ARGS = \
-  --exclude aclocal.m4 \
-  --exclude autom4te.cache \
-  --exclude bin/ \
-  --exclude configure \
-  --exclude include/config.h.in \
-  --exclude Makefile.in \
-  --exclude m4/libtool.m4 \
-  --exclude m4/ltoptions.m4 \
-  --exclude m4/ltsugar.m4 \
-  --exclude m4/ltversion.m4 \
-  --exclude m4/lt~obsolete.m4 \
-
 VKD3D_CONFIGURE_ARGS = \
   --enable-silent-rules \
   --disable-doxygen-doc \
