diff --git a/tools/winebuild/import.c b/tools/winebuild/import.c
index 9325837b7490910ae999a05253e39e457e6e87e9..baac8786a98cac9e5400ff8a77013807b4ea75da 100644
--- a/tools/winebuild/import.c
+++ b/tools/winebuild/import.c
@@ -1439,7 +1439,8 @@ void output_static_lib( const char *output_name, struct strarray files, int crea

     if (!create || target.platform != PLATFORM_WINDOWS)
     {
-        args = find_tool( "ar", NULL );
+        static const char * const commands[] = { "gcc-ar", "ar", NULL };
+        args = find_tool( "ar", commands );
         strarray_add( &args, create ? "rc" : "r" );
         strarray_add( &args, output_name );
     }
@@ -1457,7 +1458,8 @@ void output_static_lib( const char *output_name, struct strarray files, int crea

     if (target.platform != PLATFORM_WINDOWS)
     {
-        struct strarray ranlib = find_tool( "ranlib", NULL );
+        static const char * const commands[] = { "gcc-ranlib", "ranlib", NULL };
+        struct strarray ranlib = find_tool( "ranlib", commands );
         strarray_add( &ranlib, output_name );
         spawn( ranlib );
     }
diff --git a/tools/winebuild/utils.c b/tools/winebuild/utils.c
index 9f51a60c01ffd3886fd9c1d7e8e0f31b2a4a2d53..91d83fd6ba4c811cb0df4b4149db2681005c9bb4 100644
--- a/tools/winebuild/utils.c
+++ b/tools/winebuild/utils.c
@@ -389,7 +389,7 @@ const char *get_nm_command(void)
 {
     if (!nm_command.count)
     {
-        static const char * const commands[] = { "nm", "gnm", NULL };
+        static const char * const commands[] = { "gcc-nm", "nm", "gnm", NULL };
         nm_command = find_tool( "nm", commands );
     }
     if (nm_command.count > 1)
diff --git a/tools/winegcc/winegcc.c b/tools/winegcc/winegcc.c
index 56fe179c92bf23555e7e15b44c9da97b3d88dbd2..d8e13c80bbbfaa30dfaf406e85d4c8966226bfb4 100644
--- a/tools/winegcc/winegcc.c
+++ b/tools/winegcc/winegcc.c
@@ -1633,6 +1633,13 @@ int main(int argc, char **argv)
                         opts.pic = 1;
                     else if (!strcmp("-fno-PIC", opts.args.str[i]) || !strcmp("-fno-pic", opts.args.str[i]))
                         opts.pic = 0;
+                    /* LTO options are available for both the compiler and the linker */
+                    else if ((!strncmp("-flto", opts.args.str[i], 5) && (!opts.args.str[i][5] || opts.args.str[i][5] == '=')) || !strcmp("-fno-lto", opts.args.str[i]))
+                        raw_linker_arg = 1;
+                    else if (!strncmp("-flto-partition=", opts.args.str[i], 16))
+                        raw_linker_arg = 1;
+                    else if (!strcmp("-fuse-linker-plugin", opts.args.str[i]) || !strcmp("-fno-use-linker-plugin", opts.args.str[i]))
+                        raw_linker_arg = 1;
 		    break;
                 case 'i':
                     if (!strcmp( "-isysroot", opts.args.str[i] )) opts.isysroot = opts.args.str[i + 1];
