--- glibc-2.25/math/Makefile~	2017-02-05 15:28:43.000000000 +0000
+++ glibc-2.25/math/Makefile	2017-07-29 13:52:56.181213874 +0000
@@ -21,6 +21,8 @@
 
 include ../Makeconfig
 
+CFLAGS-.o += -fno-stack-protector  -falign-functions=32
+
 # Installed header files.
 headers		:= math.h bits/mathcalls.h \
 		   fpu_control.h complex.h bits/cmathcalls.h fenv.h \
--- glibc-2.25/crypt/Makefile~	2017-02-05 15:28:43.000000000 +0000
+++ glibc-2.25/crypt/Makefile	2017-07-30 18:45:58.396204712 +0000
@@ -22,6 +22,8 @@
 
 include ../Makeconfig
 
+CFLAGS-.o += -flto -ffat-lto-objects -fno-stack-protector
+
 headers := crypt.h
 
 extra-libs := libcrypt
