--- a/CMakeLists.txt	2025-06-05 19:35:35.747470966 +0200
+++ b/CMakeLists.txt	2025-06-05 19:39:22.115215750 +0200
@@ -105,7 +105,7 @@ set(CMAKE_MESSAGE_CONTEXT "")
 message(STATUS "Configuring nlohmann_json")
 set(CMAKE_MESSAGE_CONTEXT "nlohmann_json")
 if (USE_EXTERNAL_JSON)
-    find_package(nlohmann_json 3.9 REQUIRED)
+    find_package(nlohmann_json 3.9 REQUIRED)
 else()
     CPMAddPackage("gh:nlohmann/json@3.11.3")
 endif()
@@ -117,15 +117,15 @@ set(CMAKE_MESSAGE_CONTEXT "fmtlib")
 message(STATUS "Configuring fmtlib")
 if(USE_EXTERNAL_FMTLIB)
     message("Using external fmtlib")
-    # Stand-alone build
-    find_package(fmt 8.0 REQUIRED)
+    # Stand-alone build - updated version here as well for consistency
+    find_package(fmt 11.2.0 REQUIRED)
 else()
     CPMAddPackage(
         NAME fmt
         GITHUB_REPOSITORY fmtlib/fmt
         GIT_PROGRESS TRUE
         GIT_SHALLOW TRUE
-        GIT_TAG 10.1.1
+        GIT_TAG 11.2.0
         EXCLUDE_FROM_ALL YES
     )
 endif()
@@ -160,15 +160,15 @@ message(STATUS "Configuring spdlog")
 set(SPDLOG_FMT_EXTERNAL ON CACHE INTERNAL "Make spdlog use external fmtlib" FORCE)
 if(USE_EXTERNAL_SPDLOG)
     message("Using external spdlog")
-    # Stand-alone build
-    find_package(spdlog 1.9 REQUIRED)
+    # Stand-alone build - updated version here as well for consistency
+    find_package(spdlog 1.15.3 REQUIRED)
 else()
     CPMAddPackage(
         NAME spdlog
         GITHUB_REPOSITORY gabime/spdlog
         GIT_PROGRESS TRUE
         GIT_SHALLOW TRUE
-        GIT_TAG v1.12.0
+        GIT_TAG v1.15.3
         EXCLUDE_FROM_ALL YES)
 endif()
 target_link_libraries(ananicy-cpp PRIVATE project_options spdlog::spdlog fmt::fmt rt)
