--- a/CMakeLists.txt	2025-06-05 19:35:35.747470966 +0200
+++ b/CMakeLists.txt	2025-06-05 19:39:22.115215750 +0200
@@ -118,14 +118,14 @@ message(STATUS "Configuring fmtlib")
 if(USE_EXTERNAL_FMTLIB)
     message("Using external fmtlib")
     # Stand-alone build
-    find_package(fmt 8.0 REQUIRED)
+    find_package(fmt 12.1.0 REQUIRED)
 else()
     CPMAddPackage(
         NAME fmt
         GITHUB_REPOSITORY fmtlib/fmt
         GIT_PROGRESS TRUE
         GIT_SHALLOW TRUE
-        GIT_TAG 10.2.1
+        GIT_TAG 12.1.0
         EXCLUDE_FROM_ALL YES
     )
 endif()
@@ -168,7 +168,7 @@ else()
         GITHUB_REPOSITORY gabime/spdlog
         GIT_PROGRESS TRUE
         GIT_SHALLOW TRUE
-        GIT_TAG v1.14.0
+        GIT_TAG v1.16.0
         EXCLUDE_FROM_ALL YES)
 endif()
 target_link_libraries(ananicy-cpp PRIVATE project_options spdlog::spdlog fmt::fmt rt)
