--- a/pgo-bolt.sh	2023-06-07 15:04:18.565392328 +0200
+++ b/pgo-bolt.sh	2023-06-07 15:00:49.956838379 +0200
@@ -99,7 +99,12 @@ optimize_binary() {
         -dyno-stats \
         -icf=1 \
         -use-gnu-stack \
-        -plt=hot || die "Could not optimize binary"
+        -peepholes=all \
+        -elim-link-veneers \
+    -group-stubs -align-blocks -sctc-mode=heuristic -jump-tables=aggressive -simplify-conditional-tail-calls -simplify-rodata-loads \
+    -align-macro-fusion=all -eliminate-unreachable -tail-duplication=cache -indirect-call-promotion=all -icp-eliminate-loads \
+    -hot-data -x86-strip-redundant-address-size -lite=false \
+    -plt=hot || die "Could not optimize binary"
 
     echo "==== Done optimizing ${original_binary} ===="
 }
