diff --git a/src/Makefile b/src/Makefile
index 6161de3..82be56d 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -26,7 +26,7 @@ LLC ?= llc
 LLVM_STRIP ?= llvm-strip
 BPFTOOL ?= bpftool
 BPF_INCLUDE := /usr/include
-BPF_CFLAGS := -g -fno-stack-protector -Wall
+BPF_CFLAGS := -g -fno-stack-protector -std=gnu18 -Wall
 NL_INCLUDE := /usr/include/libnl3
 INCLUDES := -I../include -I$(BPF_INCLUDE) -I$(NL_INCLUDE) -I../include/uapi

@@ -55,7 +55,7 @@ endif
 VERSION = 0.1.3
 VERSION_SCRIPT  := libbpftune.map

-CFLAGS = -fPIC -Wall -Wextra -march=native -g -I../include -std=c99
+CFLAGS = -fPIC -Wall -Wextra -O3 -march=native -maes -fcf-protection=none -mharden-sls=none -flto -g -I../include -std=gnu18

 CFLAGS += -DBPFTUNE_VERSION='"$(BPFTUNE_VERSION)"' $(INCLUDES)

diff --git a/sample_tuner/Makefile b/sample_tuner/Makefile
index 6161de3..82be56d 100644
--- a/sample_tuner/Makefile
+++ b/sample_tuner/Makefile
@@ -37,7 +37,7 @@ installprefix = $(DESTDIR)/$(prefix)

 INSTALLPATH = $(installprefix)

-CFLAGS = -fPIC -Wall -Wextra -march=native -g -I../include -std=c99
+CFLAGS = -fPIC -Wall -Wextra -march=native -mtune=native -maes -g -I../include -std=gnu18

 CFLAGS += -DBPFTUNE_VERSION='"$(BPFTUNE_VERSION)"' $(INCLUDES)

@@ -88,12 +88,12 @@ $(TUNER_LIBS): $(BPF_SKELS) $(TUNER_OBJS
 	$(QUIET_GEN)$(BPFTOOL) gen skeleton $< > $@

 $(BPF_OBJS): $(patsubst %.o,%.c,$(BPF_OBJS)) ../include/bpftune/bpftune.bpf.h
-	$(CLANG) -g -D__TARGET_ARCH_$(SRCARCH) -O2 -target bpf		\
+	$(CLANG) -g -D__TARGET_ARCH_$(SRCARCH) -O3 -march=native -mtune=native -maes -target bpf		\
 		$(INCLUDES) -c $(patsubst %.o,%.c,$(@)) -o $(@);
-	$(CLANG) -g -D__TARGET_ARCH_$(SRCARCH) -DBPFTUNE_LEGACY -O2 -target bpf \
+	$(CLANG) -g -D__TARGET_ARCH_$(SRCARCH) -DBPFTUNE_LEGACY -O3 -march=native -mtune=native -maes -target bpf \
 		$(INCLUDES) -c $(patsubst %.o,%.c,$(@)) \
 		-o $(patsubst %.o,%.legacy.o,$(@));
-	$(CLANG) -g -D__TARGET_ARCH_$(SRCARCH) -DBPFTUNE_NOBTF -DBPFTUNE_LEGACY -O2 -target bpf \
+	$(CLANG) -g -D__TARGET_ARCH_$(SRCARCH) -DBPFTUNE_NOBTF -DBPFTUNE_LEGACY -O3 -march=native -mtune=native -maes -target bpf \
 		$(INCLUDES) -c $(patsubst %.o,%.c,$(@)) \
 		-o $(patsubst %.o,%.nobtf.o,$(@));
 
--- a/test/strategy/Makefile
+++ b/test/strategy/Makefile
@@ -38,7 +38,7 @@ installprefix = $(DESTDIR)/$(prefix)

 INSTALLPATH = $(installprefix)

-CFLAGS = -fPIC -Wall -Wextra -march=native -g -I../include -std=c99
+CFLAGS = -fPIC -Wall -Wextra -march=native -mtune=native -maes -g -I../include -std=gnu18

 CFLAGS += -DBPFTUNE_VERSION='"$(BPFTUNE_VERSION)"' $(INCLUDES)

@@ -87,16 +87,16 @@ $(TUNER_LIBS): $(BPF_SKELS) $(TUNER_OBJS
 $(TUNER_OBJS): $(BPF_SKELS) $(LEGACY_BPF_SKELS) $(NOBTF_BPF_SKELS)

 $(BPF_OBJS): $(patsubst %.o,%.c,$(BPF_OBJS)) ../../include/bpftune/bpftune.bpf.h
-	$(CLANG) $(BPF_CFLAGS) -D__TARGET_ARCH_$(SRCARCH) -O2 -target bpf \
+	$(CLANG) $(BPF_CFLAGS) -D__TARGET_ARCH_$(SRCARCH) -O3 -target bpf \
 		$(INCLUDES) -c $(patsubst %.o,%.c,$(@)) -o $(@)

 $(LEGACY_BPF_OBJS): $(patsubst %.legacy.o,%.c,$(LEGACY_BPF_OBJS)) ../../include/bpftune/bpftune.bpf.h
-	$(CLANG) $(BPF_CFLAGS) -D__TARGET_ARCH_$(SRCARCH) -DBPFTUNE_LEGACY -O2 -target bpf \
+	$(CLANG) $(BPF_CFLAGS) -D__TARGET_ARCH_$(SRCARCH) -DBPFTUNE_LEGACY -O3 -target bpf \
 		$(INCLUDES) -c $(patsubst %.legacy.o,%.c,$(@)) \
 		-o $(@)

 $(NOBTF_BPF_OBJS): $(patsubst %.nobtf.o,%.c,$(NOBTF_BPF_OBJS)) ../../include/bpftune/bpftune.bpf.h
-	$(CLANG) $(BPF_CFLAGS) -D__TARGET_ARCH_$(SRCARCH) -DBPFTUNE_NOBTF -O2 -target bpf \
+	$(CLANG) $(BPF_CFLAGS) -D__TARGET_ARCH_$(SRCARCH) -DBPFTUNE_NOBTF -O3 -target bpf \
 		$(INCLUDES) -c $(patsubst %.nobtf.o,%.c,$(@)) \
 		-o $(@)
