diff --git a/Makefile b/Makefile
index 6f23d53..cee6e31 100644
--- a/Makefile
+++ b/Makefile
@@ -119,7 +119,7 @@ $(OBJDIR)/.build-libmspack-dist: | $(OBJDIR)/libmspack
 	autoreconf -vfi && \
 	./configure --prefix=$(BASEDIR) --libdir=$(LIBDIR) --disable-static && \
 	sed -i -e 's/ -shared / -Wl,-O1,--as-needed\0/g' libtool && \
-	make
+	make libmspack.la
 	touch $(@)
 
 .PHONY: libmspack-dist
@@ -129,10 +129,10 @@ libmspack-dist: $(OBJDIR)/.build-libmspack-dist
 libmspack-install: libmspack-dist
 	$(info :: Installing libmspack )
 	cd $(OBJDIR)/libmspack/libmspack && \
-	make DESTDIR=$(TARGET_DIR) install
-	rm -r $(TARGET_DIR)$(BASEDIR)/include
-	rm -r $(TARGET_DIR)$(LIBDIR)/pkgconfig
-	rm    $(TARGET_DIR)$(LIBDIR)/libmspack.la
+	make DESTDIR=$(TARGET_DIR) install-libLTLIBRARIES
+	rm -rf $(TARGET_DIR)$(BASEDIR)/include
+	rm -rf $(TARGET_DIR)$(LIBDIR)/pkgconfig
+	rm -f  $(TARGET_DIR)$(LIBDIR)/libmspack.la
 
 #
 # unzip
