--- xwayland-window.h.orig	2025-10-02 15:42:48.874896417 +0200
+++ xwayland-window.h	2025-10-02 15:43:46.905645638 +0200
@@ -95,9 +95,9 @@ struct xwl_window {
     struct wl_output *wl_output_fullscreen;
     struct xorg_list xwl_output_list;
     struct xorg_list frame_callback_list;
-#ifdef XWL_HAS_LIBDECOR
+    #ifdef XWL_HAS_LIBDECOR
     struct libdecor_frame *libdecor_frame;
-#endif
+    #endif
     struct xwayland_surface_v1 *xwayland_surface;
     struct xwl_dmabuf_feedback feedback;
     /* If TRUE, the window buffer format supports scanout with implicit modifier */
@@ -135,6 +135,7 @@ int xwl_config_notify(WindowPtr window,
                       int x, int y,
                       int width, int height, int bw,
                       WindowPtr sib);
+void xwl_reparent_window(WindowPtr window, WindowPtr prior_parent);
 void xwl_resize_window(WindowPtr window,
                        int x, int y,
                        unsigned int width, unsigned int height,
