From 530c8ba03880afd93dee3dd3f24969bd0999fe8c Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Mon, 12 Dec 2022 21:00:37 +0100
Subject: [PATCH] Makefile-GCC

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 Makefile | 30 +++++++++++++++---------------
 1 file changed, 15 insertions(+), 15 deletions(-)

diff --git a/Makefile b/Makefile
index 006680a6f13d..bea94e5ce370 100644
--- a/Makefile
+++ b/Makefile
@@ -452,10 +452,10 @@ HOSTRUSTC = rustc
 HOSTPKG_CONFIG	= pkg-config
 
 KBUILD_USERHOSTCFLAGS := -Wall -Wmissing-prototypes -Wstrict-prototypes \
-			 -O2 -fomit-frame-pointer -std=gnu11 \
+			 -O3 -march=native -mtune=native -maes -fno-semantic-interposition -falign-functions=32 -floop-nest-optimize -fgraphite-identity -fdevirtualize-at-ltrans -mprefer-vector-width=256 -fipa-pta -flive-range-shrinkage -fno-math-errno -fno-trapping-math -malign-data=cacheline -fomit-frame-pointer -std=gnu18 \
 			 -Wdeclaration-after-statement
-KBUILD_USERCFLAGS  := $(KBUILD_USERHOSTCFLAGS) $(USERCFLAGS)
-KBUILD_USERLDFLAGS := $(USERLDFLAGS)
+KBUILD_USERCFLAGS  := -O3 -march=native -mtune=native -maes -fno-semantic-interposition -falign-functions=32 -floop-nest-optimize -fgraphite-identity -fdevirtualize-at-ltrans -mprefer-vector-width=256 -fipa-pta -flive-range-shrinkage -fno-math-errno -fno-trapping-math -malign-data=cacheline -fomit-frame-pointer -std=gnu18 $(KBUILD_USERHOSTCFLAGS) $(USERCFLAGS)
+KBUILD_USERLDFLAGS := -Wl,-O3,--as-needed,-Bsymbolic-functions $(USERLDFLAGS)
 
 # These flags apply to all Rust code in the tree, including the kernel and
 # host programs.
@@ -473,12 +473,12 @@ export rust_common_flags := --edition=2021 \
 			    -Dclippy::needless_continue \
 			    -Wclippy::dbg_macro
 
-KBUILD_HOSTCFLAGS   := $(KBUILD_USERHOSTCFLAGS) $(HOST_LFS_CFLAGS) $(HOSTCFLAGS)
-KBUILD_HOSTCXXFLAGS := -Wall -O2 $(HOST_LFS_CFLAGS) $(HOSTCXXFLAGS)
+KBUILD_HOSTCFLAGS   := -O3 -march=native -mtune=native -maes -fno-semantic-interposition -falign-functions=32 -floop-nest-optimize -fgraphite-identity -fdevirtualize-at-ltrans -mprefer-vector-width=256 -fipa-pta -flive-range-shrinkage -fno-math-errno -fno-trapping-math -malign-data=cacheline -fomit-frame-pointer -std=gnu18 $(KBUILD_USERHOSTCFLAGS) $(HOST_LFS_CFLAGS) $(HOSTCFLAGS)
+KBUILD_HOSTCXXFLAGS := -O3 -march=native -mtune=native -maes -fno-semantic-interposition -falign-functions=32 -floop-nest-optimize -fgraphite-identity -fdevirtualize-at-ltrans -mprefer-vector-width=256 -fipa-pta -flive-range-shrinkage -fno-math-errno -fno-trapping-math -malign-data=cacheline -fomit-frame-pointer -std=gnu++17 $(HOST_LFS_CFLAGS) $(HOSTCXXFLAGS)
 KBUILD_HOSTRUSTFLAGS := $(rust_common_flags) -O -Cstrip=debuginfo \
 			-Zallow-features= $(HOSTRUSTFLAGS)
-KBUILD_HOSTLDFLAGS  := $(HOST_LFS_LDFLAGS) $(HOSTLDFLAGS)
-KBUILD_HOSTLDLIBS   := $(HOST_LFS_LIBS) $(HOSTLDLIBS)
+KBUILD_HOSTLDFLAGS  := -Wl,-O3,--as-needed,-Bsymbolic-functions $(HOST_LFS_LDFLAGS) $(HOSTLDFLAGS)
+KBUILD_HOSTLDLIBS   := -Wl,-O3,--as-needed,-Bsymbolic-functions $(HOST_LFS_LIBS) $(HOSTLDLIBS)
 
 # Make variables (CC, etc...)
 CPP		= $(CC) -E
@@ -531,14 +531,14 @@ PAHOLE_FLAGS	= $(shell PAHOLE=$(PAHOLE) $(srctree)/scripts/pahole-flags.sh)
 CHECKFLAGS     := -D__linux__ -Dlinux -D__STDC__ -Dunix -D__unix__ \
 		  -Wbitwise -Wno-return-void -Wno-unknown-attribute $(CF)
 NOSTDINC_FLAGS :=
-CFLAGS_MODULE   =
+CFLAGS_MODULE   =-O3 -march=native -fno-semantic-interposition -falign-functions=32 -floop-nest-optimize -fgraphite-identity -fdevirtualize-at-ltrans -mprefer-vector-width=256 -fipa-pta -flive-range-shrinkage -fno-math-errno -fno-trapping-math -malign-data=cacheline -fomit-frame-pointer -std=gnu18 -Wl,-O3,--as-needed,-Bsymbolic-functions
 RUSTFLAGS_MODULE =
-AFLAGS_MODULE   =
+AFLAGS_MODULE   = -D__AVX__=1 -D__AVX2__=1 -msse2avx -D__FMA__=1
 LDFLAGS_MODULE  =
-CFLAGS_KERNEL	=
+CFLAGS_KERNEL	= -march=native -fno-semantic-interposition -falign-functions=32 -floop-nest-optimize -fgraphite-identity -fdevirtualize-at-ltrans -mprefer-vector-width=256 -fipa-pta -flive-range-shrinkage -fno-math-errno -fno-trapping-math -malign-data=cacheline -fomit-frame-pointer -std=gnu18 -Wl,-O3,--as-needed,-Bsymbolic-functions
 RUSTFLAGS_KERNEL =
-AFLAGS_KERNEL	=
-export LDFLAGS_vmlinux =
+AFLAGS_KERNEL	= -D__AVX__=1 -D__AVX2__=1 -msse2avx -D__FMA__=1
+export LDFLAGS_vmlinux = -Wl,-O3,--as-needed,-Bsymbolic-functions
 
 # Use USERINCLUDE when you must reference the UAPI directories only.
 USERINCLUDE    := \
@@ -563,7 +563,7 @@ KBUILD_CFLAGS   := -Wall -Wundef -Werror=strict-prototypes -Wno-trigraphs \
 		   -fno-strict-aliasing -fno-common -fshort-wchar -fno-PIE \
 		   -Werror=implicit-function-declaration -Werror=implicit-int \
 		   -Werror=return-type -Wno-format-security \
-		   -std=gnu11
+		   -std=gnu18
 KBUILD_CPPFLAGS := -D__KERNEL__
 KBUILD_RUSTFLAGS := $(rust_common_flags) \
 		    --target=$(objtree)/rust/target.json \

-- 
2.39.0

