diff --git a/arch/x86/Makefile b/arch/x86/Makefile
--- a/arch/x86/Makefile
+++ b/arch/x86/Makefile
@@ -12,29 +12,6 @@ else
         KBUILD_DEFCONFIG := $(ARCH)_defconfig
 endif
 
-ifdef CONFIG_CC_IS_GCC
-RETPOLINE_CFLAGS	:= $(call cc-option,-mindirect-branch=thunk-extern -mindirect-branch-register)
-RETPOLINE_VDSO_CFLAGS	:= $(call cc-option,-mindirect-branch=thunk-inline -mindirect-branch-register)
-endif
-ifdef CONFIG_CC_IS_CLANG
-RETPOLINE_CFLAGS	:= -mretpoline-external-thunk
-RETPOLINE_VDSO_CFLAGS	:= -mretpoline
-endif
-RETPOLINE_CFLAGS	+= $(call cc-option,-mindirect-branch-cs-prefix)
-
-ifdef CONFIG_MITIGATION_RETHUNK
-RETHUNK_CFLAGS		:= -mfunction-return=thunk-extern
-RETHUNK_RUSTFLAGS	:= -Zfunction-return=thunk-extern
-RETPOLINE_CFLAGS	+= $(RETHUNK_CFLAGS)
-RETPOLINE_RUSTFLAGS	+= $(RETHUNK_RUSTFLAGS)
-endif
-
-export RETHUNK_CFLAGS
-export RETHUNK_RUSTFLAGS
-export RETPOLINE_CFLAGS
-export RETPOLINE_RUSTFLAGS
-export RETPOLINE_VDSO_CFLAGS
-
 # For gcc stack alignment is specified with -mpreferred-stack-boundary,
 # clang has the option -mstack-alignment for that purpose.
 ifneq ($(call cc-option, -mpreferred-stack-boundary=4),)
@@ -47,8 +24,8 @@ endif
 
 # How to compile the 16-bit code.  Note we always compile for -march=i386;
 # that way we can complain to the user if the CPU is insufficient.
-REALMODE_CFLAGS	:= -std=gnu11 -m16 -g -Os -DDISABLE_BRANCH_PROFILING -D__DISABLE_EXPORTS \
-		   -Wall -Wstrict-prototypes -march=i386 -mregparm=3 \
+REALMODE_CFLAGS	:= -std=gnu11 -m16 -O2 -march=native -DDISABLE_BRANCH_PROFILING -D__DISABLE_EXPORTS \
+		   -Wall -Wstrict-prototypes -mregparm=3 \
 		   -fno-strict-aliasing -fomit-frame-pointer -fno-pic \
 		   -mno-mmx -mno-sse $(call cc-option,-fcf-protection=none)
 
@@ -74,14 +51,14 @@ export BITS
 #
 #    https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53383
 #
-KBUILD_CFLAGS += -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -mno-avx
+KBUILD_CFLAGS += -mno-sse -mno-mmx -mno-sse2
 KBUILD_RUSTFLAGS += --target=$(objtree)/scripts/target.json
-KBUILD_RUSTFLAGS += -Ctarget-feature=-sse,-sse2,-sse3,-ssse3,-sse4.1,-sse4.2,-avx,-avx2
+KBUILD_RUSTFLAGS += -Ctarget-cpu=nativ
 
 #
 # CFLAGS for compiling floating point code inside the kernel.
 #
-CC_FLAGS_FPU := -msse -msse2
+CC_FLAGS_FPU := -msse -msse2 -mavx -mavx2
 ifdef CONFIG_CC_IS_GCC
 # Stack alignment mismatch, proceed with caution.
 # GCC < 7.1 cannot compile code using `double` and -mpreferred-stack-boundary=3
